### DISCUSSION TITLE
Pocketbase api rules are applied after record save ?

### DISCUSSION THREAD

**Discussion by neerajnigam6** [pocketbase/pocketbase#4112] [Q&A] [answered]
I love Pocketbase and created an app over it. However, I am facing a really weird problem related to API rules. 

This problem is specific to create rule, 
`org_user` collection has two foreign keys, 1 to `user` collection and the other to `org_role` collection. 
`org_role` collection has different sets of permissions and a foreign key to `organisation` collection.
here is my create a rule for `org_user` collection:

(note organisations is a single reference, it was a typo during the creation of the collection)
`@collection.org_user.role.organisations ?= @request.data.role.organisations && @collection.org_user.user ?= @request.auth.id && @collection.org_user.role.manage_roles ?= true`

Issue: 
A user who doesn't have any role with `manage_roles` permission should not be able to create an entry in `org_user` collection.
This is expected behavior and it is working fine, however, when a user tries to assign a new role where `manage_roles` permission is set to true, he is able to create this entry in DB. 
My thoughts are, this is because the new role that is going to be added to the collection has `manage_roles` permission, and it is somehow part of my query. How can I modify the API rules to achieve this behavior?

**Comment by ganigeorgiev** [comment, marked as answer]
The Create API rule runs after a "dry save" of the record (_there should be a tooltip next to the Create Rule label in the Admin UI_). The reasons for this is because API rules are translated to raw SQL statements and we need the record to be existing in the database to evaluate the rule expressions.

I'm having difficulties understanding your collections structure and relations but if your problem is that the `@collection.*` includes also your currently created entry then you can try to exclude it by adding to your rule something like `@collection.org_user.id ?!= id && ...`.

**Reply by neerajnigam6** [reply]
Thanks a ton, this is exactly what I was looking for. I wanted to evaluate the rule before the creation of the record.

**Reply by ganigeorgiev** [reply]
_I've added a [task in the roadmap](https://github.com/orgs/pocketbase/projects/2/views/1?pane=issue&itemId=49458766) to consider eventually changing the Create API rule behavior ~for v0.21 or v0.22~ in the future to run it against a temp one-off empty collection with the same fields so that it can be isolated and have a more predictable behavior._

**Reply by doceazedo** [reply]
Are there any alternatives to this in the meantime? How can I, for example, make sure a user has not already liked a post if `@request.auth.likes_via_from.post != post` will run after the dry save?

**Reply by ganigeorgiev** [reply]
@doceazedo Without details about your collections structure it is difficult to understand what do you mean by _"user has not already liked a post"_.

If you want to prevent duplicates then a better approach would be to define a UNIQUE index for your fields.

If you want to match records, then you can use the `@collection.*` filter as suggested above but the caveat is that since it executes after the creation your condition must also include the value you want to check.

Another alternative could be to create a helper view collection that will count the related records and which you can join with the `@collection.*` filter and check its count field (it would be +1).

**Reply by doceazedo** [reply]
I see, a UNIQUE index would be sufficient here. I guess I wasn't aware about them. Thanks!
